engine/src/flutter/lib/ui/painting/pixel_deferred_image_gpu_impeller.cc

```diff
@@ -0,0 +1,133 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/painting/pixel_deferred_image_gpu_impeller.h"
+
+#include "flutter/fml/make_copyable.h"
+#include "flutter/fml/trace_event.h"
+
+namespace flutter {
+
+sk_sp<PixelDeferredImageGPUImpeller> PixelDeferredImageGPUImpeller::Make(
+    sk_sp<SkImage> image,
+    fml::TaskRunnerAffineWeakPtr<SnapshotDelegate> snapshot_delegate,
+    fml::RefPtr<fml::TaskRunner> raster_task_runner) {
+  return sk_sp<PixelDeferredImageGPUImpeller>(new PixelDeferredImageGPUImpeller(
+      PixelDeferredImageGPUImpeller::ImageWrapper::Make(
+          std::move(image), std::move(snapshot_delegate),
+          std::move(raster_task_runner))));
+}
+
+PixelDeferredImageGPUImpeller::PixelDeferredImageGPUImpeller(
+    std::shared_ptr<ImageWrapper> wrapper)
+    : wrapper_(std::move(wrapper)) {}
+
+PixelDeferredImageGPUImpeller::~PixelDeferredImageGPUImpeller() = default;
+
+sk_sp<SkImage> PixelDeferredImageGPUImpeller::skia_image() const {
+  return nullptr;
+}
+
+std::shared_ptr<impeller::Texture>
+PixelDeferredImageGPUImpeller::impeller_texture() const {
+  if (!wrapper_) {
+    return nullptr;
+  }
+  return wrapper_->texture();
+}
+
+bool PixelDeferredImageGPUImpeller::isOpaque() const {
+  return false;
+}
+
+bool PixelDeferredImageGPUImpeller::isTextureBacked() const {
+  return wrapper_ && wrapper_->isTextureBacked();
+}
+
+bool PixelDeferredImageGPUImpeller::isUIThreadSafe() const {
+  return true;
+}
+
+DlISize PixelDeferredImageGPUImpeller::GetSize() const {
+  return wrapper_ ? wrapper_->size() : DlISize();
+}
+
+size_t PixelDeferredImageGPUImpeller::GetApproximateByteSize() const {
+  auto size = sizeof(PixelDeferredImageGPUImpeller);
+  if (wrapper_) {
+    if (wrapper_->texture()) {
+      size += wrapper_->texture()
+                  ->GetTextureDescriptor()
+                  .GetByteSizeOfBaseMipLevel();
+    } else {
+      size += wrapper_->size().Area() * 4;
+    }
```

Comment:
So the reported size might change asynchronously as the image is decoded and delivered? Is there a caller that might run into problems with that?

Reply:
This is used by the raster cache.  The callers are all about estimating size of things.  This is also how DlDeferredImageGPUImpeller::GetApproximateByteSize works so it should be good.